lib/ngx/re.lua

```diff
@@ -0,0 +1,154 @@
+local bit = require "bit"
+local base = require "resty.core.base"
+local core_regex = require "resty.core.regex"
+
+
+local C = ffi.C
+local sub = string.sub
+local band = bit.band
+local new_tab = base.new_tab
+local tostring = tostring
+local re_match_compile = core_regex.re_match_compile
+local destroy_compiled_regex = core_regex.destroy_compiled_regex
+
+
+local PCRE_ERROR_NOMATCH = -1
+
+
+local FLAG_DFA           = 0x02
+
+
+ffi.cdef [[
+    int ngx_http_lua_ffi_exec_regex(ngx_http_lua_regex_t *re, int flags,
+        const unsigned char *s, size_t len, int pos);
+]]
+
+
+local _M = { version = base.version }
+
+
+local function re_split_helper(subj, regex, opts, ctx)
+    local compiled, compile_once, flags = re_match_compile(regex, opts)
+    if compiled == nil then
+        -- compiled_once holds the error string
+        return nil, nil, nil, compile_once
+    end
+
+    -- exec the compiled regex
+
+    local rc
+    do
+        local pos = ctx.pos - 1
+        if pos < 0 then
+            pos = 0
+        end
+
+        rc = C.ngx_http_lua_ffi_exec_regex(compiled, flags, subj, #subj, pos)
+    end
+
+    if rc == PCRE_ERROR_NOMATCH then
+        if not compile_once then
+            destroy_compiled_regex(compiled)
+        end
+        return nil, nil, nil
+    end
+
+    if rc < 0 then
+        if not compile_once then
+            destroy_compiled_regex(compiled)
+        end
+        return nil, nil, nil, "pcre_exec() failed: " .. rc
+    end
+
+    if rc == 0 then
+        if band(flags, FLAG_DFA) == 0 then
+            return nil, nil, nil, "capture size too small"
+        end
+
+        rc = 1
+    end
+
+    local caps = compiled.captures
+    local ncaps = compiled.ncaptures
+
+    local from = caps[0] + 1
+    local to = caps[1]
+
+    if from < 0 or to < 0 then
+        return nil, nil, nil
+    end
+
+    ctx.pos = to + 1
+
+    -- retrieve the first sub-match capture if any
+
+    if ncaps > 0 and rc > 1 then
+        return from, to, sub(subj, caps[2] + 1, caps[3])
+    end
+
+    return from, to
+end
+
+
+function _M.split(subj, regex, opts, ctx, max, res)
+    -- we need to cast this to strings to avoid exceptions when they are
+    -- something else.
+    -- needed because of further calls to string.sub in this function.
+    subj = tostring(subj)
+
+    if not ctx then
+        ctx = new_tab(0, 1)
+    end
+
+    if not ctx.pos then
+        ctx.pos = 1
+    end
+
+    if not res then
+        res = new_tab(1, 0)
+    elseif type(res) ~= "table" then
+        return error("res is not a table", 2)
+    end
+
+    local sub_idx = ctx.pos
+    local res_idx = 0
+    local count = 0
+
+    while true do
+        local from, to, capture, err = re_split_helper(subj, regex, opts, ctx)
+        if err then
+            return nil, err
+        end
+
+        if not from then
+            res[res_idx + 1] = sub(subj, sub_idx, #subj)
+            res[res_idx + 2] = nil -- trailing nil for non-cleared res tables
```

Comment:
Doing so would be a mistake, since instead of:
```lua
local res, err = ngx_re.split("a,b,c,d,e", ",", nil, nil, 3)
-- res is {"a", "b", "c"}
```

We would have:
```lua
local res, err = ngx_re.split("a,b,c,d,e", ",", nil, nil, 3)
-- res is {"a", "b", "c", "d,e"}
```